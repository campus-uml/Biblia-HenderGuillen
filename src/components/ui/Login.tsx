import React, { useState } from "react";
import {
  loginWithEmailAndPassword,
  registerWithEmailAndPassword,
  googleProvider,
  githubProvider,
  auth,
} from "./firebase";
import { signInWithPopup } from "firebase/auth";
import { Button } from "./button";
import { Card, CardContent, CardHeader, CardTitle } from "./card";
import { FaGoogle, FaGithub, FaBook } from "react-icons/fa";

interface LoginProps {
  onLogin: (user: any) => void;
}

const Login: React.FC<LoginProps> = ({ onLogin }) => {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState<string | null>(null);

  const handleLogin = async () => {
    try {
      const userCredential = await loginWithEmailAndPassword(email, password);
      console.log("Inicio de sesión exitoso");
      onLogin(userCredential.user);
    } catch (err: any) {
      setError(err.message);
    }
  };

  const handleRegister = async () => {
    try {
      const userCredential = await registerWithEmailAndPassword(email, password);
      console.log("Registro exitoso");
      onLogin(userCredential.user);
    } catch (err: any) {
      setError(err.message);
    }
  };

  const signInWithProvider = async (provider: any) => {
    try {
      const result = await signInWithPopup(auth, provider);
      console.log("Inicio de sesión con proveedor exitoso");
      onLogin(result.user);
    } catch (err: any) {
      console.error("Error al iniciar sesión con proveedor:", err);
      setError(err.message);
    }
  };

  return (
    <div className="min-h-screen flex items-center justify-center bg-gray-100">
      <Card className="w-full max-w-md min-h-[500px] flex flex-col justify-between bg-white p-6 rounded-lg shadow-xl">
        <CardHeader>
          <CardTitle className="text-center text-2xl font-bold">
            <div className="flex items-center justify-center space-x-2">
              <FaBook className="text-3xl text-yellow-600" />
              <span className="text-gray-700">Bienvenido a la Biblia Reina Valera</span>
            </div>
          </CardTitle>
        </CardHeader>
        <CardContent className="flex flex-col justify-center grow space-y-4">
          <div className="text-center text-lg text-gray-600 font-semibold">
            <p>Inicia sesión para continuar</p>
          </div>

          {error && (
            <div className="text-red-600 text-center bg-red-100 p-2 rounded-md">
              <p>{error}</p>
            </div>
          )}

          <input
            type="email"
            placeholder="Correo electrónico"
            value={email}
            onChange={(e) => setEmail(e.target.value)}
            className="w-full px-4 py-2 border rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-500"
          />
          <input
            type="password"
            placeholder="Contraseña"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
            className="w-full px-4 py-2 border rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-500"
          />

          <Button
            onClick={handleLogin}
            className="w-full bg-blue-600 text-white font-semibold py-2 rounded-lg hover:bg-blue-500 transition"
          >
            Iniciar sesión
          </Button>

          <Button
            onClick={handleRegister}
            className="w-full bg-green-600 text-white font-semibold py-2 rounded-lg hover:bg-green-500 transition"
          >
            Registrarse
          </Button>

          <div className="flex items-center gap-2 my-4">
            <hr className="flex-grow border-gray-300" />
            <span className="text-gray-500 font-medium whitespace-nowrap">O inicia sesión con</span>
            <hr className="flex-grow border-gray-300" />
          </div>

          <Button
            onClick={() => signInWithProvider(googleProvider)}
            className="w-full bg-red-600 text-white font-semibold py-2 rounded-lg hover:bg-red-500 flex items-center justify-center space-x-2 transition"
          >
            <FaGoogle className="text-xl" />
            <span>Ingresar con Google</span>
          </Button>

          <Button
            onClick={() => signInWithProvider(githubProvider)}
            className="w-full bg-gray-900 text-white font-semibold py-2 rounded-lg hover:bg-gray-700 flex items-center justify-center space-x-2 transition"
          >
            <FaGithub className="text-xl" />
            <span>Ingresar con GitHub</span>
          </Button>
        </CardContent>
      </Card>
    </div>
  );
};

export default Login;
