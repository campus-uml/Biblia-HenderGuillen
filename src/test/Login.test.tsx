import { describe, it, expect, vi } from "vitest";
import { render, fireEvent, screen } from "@testing-library/react";
import Login from "../components/ui/Login";

describe("Testing de Login.tsx", () => {
  it("Renderiza correctamente el componente", () => {
    const mockOnLogin = vi.fn();
    const { baseElement } = render(<Login onLogin={mockOnLogin} />);
    expect(baseElement).toMatchSnapshot();
  });

  

  

  it("Debe mostrar un mensaje de error si las credenciales son incorrectas", async () => {
    const mockOnLogin = vi.fn();
    render(<Login onLogin={mockOnLogin} />);

    const loginButton = screen.getByText(/Iniciar sesi√≥n/i);
    fireEvent.click(loginButton);

    const errorMessage = await screen.findByText(/Error/i);
    expect(errorMessage).toBeTruthy();
  });
});
