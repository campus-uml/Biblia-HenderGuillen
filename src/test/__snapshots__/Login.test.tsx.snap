// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing de Login.tsx > Renderiza correctamente el componente 1`] = `
<body>
  <div>
    <div
      class="min-h-screen flex items-center justify-center bg-gray-100"
    >
      <div
        class="border text-card-foreground w-full max-w-md min-h-[500px] flex flex-col justify-between bg-white p-6 rounded-lg shadow-xl"
      >
        <div
          class="flex flex-col space-y-1.5 p-6"
        >
          <div
            class="tracking-tight text-center text-2xl font-bold"
          >
            <div
              class="flex items-center justify-center space-x-2"
            >
              <svg
                class="text-3xl text-yellow-600"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 448 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M448 360V24c0-13.3-10.7-24-24-24H96C43 0 0 43 0 96v320c0 53 43 96 96 96h328c13.3 0 24-10.7 24-24v-16c0-7.5-3.5-14.3-8.9-18.7-4.2-15.4-4.2-59.3 0-74.7 5.4-4.3 8.9-11.1 8.9-18.6zM128 134c0-3.3 2.7-6 6-6h212c3.3 0 6 2.7 6 6v20c0 3.3-2.7 6-6 6H134c-3.3 0-6-2.7-6-6v-20zm0 64c0-3.3 2.7-6 6-6h212c3.3 0 6 2.7 6 6v20c0 3.3-2.7 6-6 6H134c-3.3 0-6-2.7-6-6v-20zm253.4 250H96c-17.7 0-32-14.3-32-32 0-17.6 14.4-32 32-32h285.4c-1.9 17.1-1.9 46.9 0 64z"
                />
              </svg>
              <span
                class="text-gray-700"
              >
                Bienvenido a la Biblia Reina Valera
              </span>
            </div>
          </div>
        </div>
        <div
          class="p-6 pt-0 flex flex-col justify-center grow space-y-4"
        >
          <div
            class="text-center text-lg text-gray-600 font-semibold"
          >
            <p>
              Inicia sesión para continuar
            </p>
          </div>
          <input
            class="w-full px-4 py-2 border rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-500"
            placeholder="Correo electrónico"
            type="email"
            value=""
          />
          <input
            class="w-full px-4 py-2 border rounded-lg focus:outline-none focus:ring-2 focus:ring-blue-500"
            placeholder="Contraseña"
            type="password"
            value=""
          />
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 shadow h-9 px-4 w-full bg-blue-600 text-white font-semibold py-2 rounded-lg hover:bg-blue-500 transition"
          >
            Iniciar sesión
          </button>
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 shadow h-9 px-4 w-full bg-green-600 text-white font-semibold py-2 rounded-lg hover:bg-green-500 transition"
          >
            Registrarse
          </button>
          <div
            class="flex items-center gap-2 my-4"
          >
            <hr
              class="flex-grow border-gray-300"
            />
            <span
              class="text-gray-500 font-medium whitespace-nowrap"
            >
              O inicia sesión con
            </span>
            <hr
              class="flex-grow border-gray-300"
            />
          </div>
          <button
            class="gap-2 whitespace-nowrap text-sm focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 shadow h-9 px-4 w-full bg-red-600 text-white font-semibold py-2 rounded-lg hover:bg-red-500 flex items-center justify-center space-x-2 transition"
          >
            <svg
              class="text-xl"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 488 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M488 261.8C488 403.3 391.1 504 248 504 110.8 504 0 393.2 0 256S110.8 8 248 8c66.8 0 123 24.5 166.3 64.9l-67.5 64.9C258.5 52.6 94.3 116.6 94.3 256c0 86.5 69.1 156.6 153.7 156.6 98.2 0 135-70.4 140.8-106.9H248v-85.3h236.1c2.3 12.7 3.9 24.9 3.9 41.4z"
              />
            </svg>
            <span>
              Ingresar con Google
            </span>
          </button>
          <button
            class="gap-2 whitespace-nowrap text-sm focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 shadow h-9 px-4 w-full bg-gray-900 text-white font-semibold py-2 rounded-lg hover:bg-gray-700 flex items-center justify-center space-x-2 transition"
          >
            <svg
              class="text-xl"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 496 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
              />
            </svg>
            <span>
              Ingresar con GitHub
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
